### PR TITLE
bats: add tablespaces to finalize.bats for testing coverage

### PR DESCRIPTION
**Pull request by dkrieger-pvtl**
Create a tablespace in a 5X cluster for upgrade. Note that, since pg_upgrade does fails to upgrade tablespaces in 6-6 mode, we only add a tablespace in GPDB5.

As a side effect of the changes made, we have a minor additional prefactor comment:  use correct psql binary in get_segment_configration.

The code change in this PR was straightforward, but much of the work was in realizing that the GPUPGRADE_TARGET rpm in our CI pipeline was outdated.  Hence, it did not contain a fix to 6X_STABLE that caused this addition to fail on the CI until the rpm version was updated.  Both prod and dev have been updated to use `greenplum-db-6.9.0-rhel7-x86_64.rpm` as their target cluster rpm.

### CODE REVIEW

File: test/finalize.bats [hunk 1]
```diff
@@ -143,6 +157,12 @@ setup_state_dir() {
     export GPUPGRADE_HOME="${STATE_DIR}/gpupgrade"
 }
 
+cleanup_state_dir() {
+    if [ -n "$STATE_DIR" ]; then
+        rm -r "$STATE_DIR"
+    fi
+}
+
```
Comment (kalensk) [inline test/finalize.bats:165]: Any particular reason for this code motion in this particular commit? Seems unrelated or not needed for directly testing tablesapces?

File: test/finalize.bats [hunk 2]
```diff
@@ -170,3 +190,52 @@ validate_data_directories() {
...
+    "${GPHOME_SOURCE}"/bin/psql -d postgres -c "DROP TABLE IF EXISTS batsTable;"
+    "${GPHOME_SOURCE}"/bin/psql -d postgres -c "DROP TABLESPACE IF EXISTS batsTbsp;"
+    "${GPHOME_SOURCE}"/bin/psql -d postgres -c "DROP FILESPACE IF EXISTS batsFS;"
+}
+
+check_tablespace_data() {
+    local row_count
+    row_count=$("$GPHOME_TARGET"/bin/psql -d postgres -Atc "SELECT COUNT(*) FROM batsTable;")
+    if (( row_count != 100 )); then
+        fail "Table batsTable only has $row_count rows; should have 100"
```
Comment (kalensk) [inline test/finalize.bats:239]: I think the error message can give a bit more context in that its trying to verify tablesaces:
`failed verifying tablespaces. batsTable got $rows want 100`

File: test/finalize.bats [hunk 3]
```diff
@@ -170,3 +190,52 @@ validate_data_directories() {
             [ ! -f "${datadir}/${marker_file}" ] || fail "unexpected ${marker_file} marker file in target datadir: $datadir"
         done
 }
+
+
+
+# This is 5X-only due to a bug in pg_upgrade for 6-6
```
Comment (kalensk) [inline test/finalize.bats:196]: Do we have any further details on the bug that would be useful breadcrumbs to expand this comment?

File: test/finalize.bats [hunk 4]
```diff
@@ -170,3 +190,52 @@ validate_data_directories() {
...
+				$(hostname):2:${TABLESPACE_ROOT}/p1/demoDataDir0
+				$(hostname):3:${TABLESPACE_ROOT}/p2/demoDataDir1
+				$(hostname):4:${TABLESPACE_ROOT}/p3/demoDataDir2
+				$(hostname):5:${TABLESPACE_ROOT}/m1/demoDataDir0
+				$(hostname):6:${TABLESPACE_ROOT}/m2/demoDataDir1
+				$(hostname):7:${TABLESPACE_ROOT}/m3/demoDataDir2
+				$(hostname):8:${TABLESPACE_ROOT}/m/standby
+EOF
+
+    gpfilespace --config "${TABLESPACE_CONFIG}"
```
Comment (kalensk) [inline test/finalize.bats:220]: I guess this doesn't need a subshell that sources GPHOME_SOURCE...Does it make sense to fully qualify the path for safety - `$GPHOME_SOURCE/bin/gpfilesapce`?

File: test/finalize.bats [hunk 5]
```diff
@@ -170,3 +190,52 @@ validate_data_directories() {
...
+    "${GPHOME_SOURCE}"/bin/psql -d postgres -c "CREATE TABLE batsTable(a int) TABLESPACE batsTbsp;"
+    "${GPHOME_SOURCE}"/bin/psql -d postgres -c "INSERT INTO batsTable SELECT i from generate_series(1,100)i;"
+}
+
+# This is 5X-only
+delete_tablespace_data() {
+    "${GPHOME_SOURCE}"/bin/psql -d postgres -c "DROP TABLE IF EXISTS batsTable;"
+    "${GPHOME_SOURCE}"/bin/psql -d postgres -c "DROP TABLESPACE IF EXISTS batsTbsp;"
+    "${GPHOME_SOURCE}"/bin/psql -d postgres -c "DROP FILESPACE IF EXISTS batsFS;"
+}
```
Comment (kalensk) [inline test/finalize.bats:233]: Might be more readable to combine these into a single statement or a single multi-line PSQL command?

File: test/finalize.bats [hunk 6]
```diff
@@ -34,6 +32,13 @@ teardown() {
     source "${GPHOME_SOURCE}/greenplum_path.sh"
     gpstart -a
 
+    # delete tablespace data added to the source cluster
+    if is_GPDB5 "${GPHOME_SOURCE}"; then
+        delete_tablespace_data
+    fi
+
+    # delete the state_dir, which also contains the tablepsace filesystem
```
Comment (bhuvnesh2703) [inline test/finalize.bats:40]: tablepsace - spelling

File: test/finalize.bats [hunk 7]
```diff
@@ -170,3 +190,52 @@ validate_data_directories() {
             [ ! -f "${datadir}/${marker_file}" ] || fail "unexpected ${marker_file} marker file in target datadir: $datadir"
         done
 }
+
+
+
+# This is 5X-only due to a bug in pg_upgrade for 6-6
+create_tablespace_with_table() {
+    # the tablespace directory will get deleted when the STATE_DIR does in teardown()
```
Comment (bhuvnesh2703) [inline test/finalize.bats:198]: ## the tablespace directory will get deleted when the STATE_DIR `is deleted` in teardown()

File: test/finalize.bats [hunk 8]
```diff
@@ -34,6 +32,13 @@ teardown() {
     source "${GPHOME_SOURCE}/greenplum_path.sh"
     gpstart -a
 
+    # delete tablespace data added to the source cluster
+    if is_GPDB5 "${GPHOME_SOURCE}"; then
+        delete_tablespace_data
+    fi
```
Comment (kalensk) [inline test/finalize.bats:38]: I think we would want to make this if statement more robust in case someone adds another test to finalize or something. That is, we only want to delete tablespaces if they have been created, and not just if its GPDB 5X.

File: test/finalize.bats [hunk 9]
```diff
@@ -170,3 +190,52 @@ validate_data_directories() {
...
         done
 }
+
+
+
+# This is 5X-only due to a bug in pg_upgrade for 6-6
+create_tablespace_with_table() {
+    # the tablespace directory will get deleted when the STATE_DIR does in teardown()
+    TABLESPACE_ROOT="${STATE_DIR}"/testfs
+    TABLESPACE_CONFIG="${TABLESPACE_ROOT}"/"fs.txt"
```
Comment (bhuvnesh2703) [inline test/finalize.bats:200]: nit: not required in quotes, "fs.txt"

File: test/finalize.bats [hunk 10]
```diff
@@ -170,3 +190,52 @@ validate_data_directories() {
...
+
+    # create the directories required to implement our filespace
+    mkdir -p "${TABLESPACE_ROOT}"
```
Comment (bhuvnesh2703) [inline test/finalize.bats:203]: since using -p don't need this mkdir, and you can combine them (but upto u)

File: test/finalize.bats (hunk 8 above)
Comment (dkrieger-pvtl) [reply test/finalize.bats:38]: I thought about this, but the `delete_tablespace_data` only drops objects `IF EXISTS` so it does no harm.  Setting a global variable seemed uglier.  This could also be added to a teardown array of functions.

File: test/finalize.bats (hunk 1 above)
Comment (dkrieger-pvtl) [reply test/finalize.bats:165]: This is new code, not code motion.  We weren't deleting the state directory(erroneously) before.

File: test/finalize.bats (hunk 3 above)
Comment (dkrieger-pvtl) [reply test/finalize.bats:196]: I avoided any more details here, because they are not relevant for this bats test.  Once 6-6 upgrade of tablespaces works, we can remove the branching on 5X in this test.

Comment (dkrieger-pvtl) [reply test/finalize.bats:196]: I could say instead
`# This is 5X-only because pg_upgrade currently does not work for 6-6`

File: test/finalize.bats (hunk 4 above)
Comment (dkrieger-pvtl) [reply test/finalize.bats:220]: Great point...updated to reference the absolute path as you suggest.  I don't think we specify, but I presumed that since the SOURCE cluster must be running upon entry that its environment was sourced at the start of the function.

File: test/finalize.bats (hunk 5 above)
Comment (dkrieger-pvtl) [reply test/finalize.bats:233]: Yes, done in both cases.  But I use a here doc to make sure each line is in a separate transactions.  See [psql -c documentation](https://www.postgresql.org/docs/9.4/app-psql.html).  Indeed, in `delete_tablespace_data`, the table needs to be dropped in a transaction prior to dropping the tablespace.

File: test/finalize.bats (hunk 4 above)
Comment (dkrieger-pvtl) [reply test/finalize.bats]: I also added source'ing GPHOME_SOURCE's greenplum_path.sh to reduce any ordering dependencies.  In other words to make sure we are calling what we are thinking of calling.
